packages/material-ui-lab/src/MasonryItem/MasonryItem.js

```diff
@@ -0,0 +1,133 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { unstable_composeClasses as composeClasses } from '@material-ui/unstyled';
+import { styled, useThemeProps } from '@material-ui/core/styles';
+import { isFragment } from 'react-is';
+import { isMuiElement } from '@material-ui/core/utils';
+import masonryItemClasses, { getMasonryItemUtilityClass } from './masonryItemClasses';
+
+const useUtilityClasses = (styleProps) => {
+  const { classes } = styleProps;
+
+  const slots = {
+    root: ['root'],
+    img: ['img'],
+    div: ['div'],
+  };
+
+  return composeClasses(slots, getMasonryItemUtilityClass, classes);
+};
+
+const MasonryItemRoot = styled('div', {
+  name: 'MuiMasonryItem',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [
+      { [`& .${masonryItemClasses.img}`]: styles.img },
+      { [`& .${masonryItemClasses.div}`]: styles.div },
+      styles.root,
+    ];
+  },
+})(() => ({
+  width: '100%',
+  [`& .${masonryItemClasses.img}`]: {
+    objectFit: 'cover',
+    width: '100%',
+  },
+  [`& .${masonryItemClasses.div}`]: {
+    width: '100%',
+  },
+}));
+
+const MasonryItem = React.forwardRef(function MasonryItem(inProps, ref) {
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonryItem',
+  });
+
+  const { children, className, component = 'div', style, ...other } = props;
+
+  const styleProps = {
+    ...props,
+    component,
+  };
+
+  const classes = useUtilityClasses(styleProps);
+
+  return (
+    <MasonryItemRoot
+      as={component}
+      className={clsx(classes.root, className)}
+      ref={ref}
+      style={style}
+      styleProps={styleProps}
+      {...other}
+    >
+      {React.Children.map(children, (child) => {
+        if (!React.isValidElement(child)) {
+          return null;
+        }
+
+        if (process.env.NODE_ENV !== 'production') {
+          if (isFragment(child)) {
+            console.error(
+              [
+                "Material-UI: The MasonryItem component doesn't accept a Fragment as a child.",
+                'Consider providing an array instead.',
+              ].join('\n'),
+            );
+          }
+        }
+
+        if (child.type === 'img' || isMuiElement(child, ['Image'])) {
```

Comment:
Let's try just to add the styles on the `div`, `img` HTML elements and avoid cloning if that's possible.